src/subdevice/configuration.rs

```diff
@@ -14,7 +14,364 @@ use crate::{
     subdevice_state::SubDeviceState,
     sync_manager_channel::{Enable, Status, SyncManagerChannel, SM_BASE_ADDRESS, SM_TYPE_ADDRESS},
 };
-use core::ops::DerefMut;
+use core::{fmt::{Debug, Display}, ops::DerefMut};
+
+/// TODO
+pub trait ConfigureSubdevicePdi {
+    /// TODO
+    type Error;
+
+    /// TODO
+    async fn configure<'a, S>(
+        &mut self,
+        subdevice: &mut SubDeviceRef<'a, S>,
+        sync_managers: &[SyncManager],
+        fmmu_usage: &[FmmuUsage],
+        direction: PdoDirection,
+        global_offset: &mut PdiOffset,
+    ) -> Result<PdiSegment, Self::Error>
+    where
+        S: DerefMut<Target = SubDevice>;
+}
+
+/// TODO
+pub enum ConfigureSubdevicePdiError<T> {
+    /// TODO
+    Callback(T),
+    /// TODO
+    Internal(Error)
+}
+
+impl<T: Display> Display for ConfigureSubdevicePdiError<T> {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        match self {
+            ConfigureSubdevicePdiError::Callback(e) => e.fmt(f),
+            ConfigureSubdevicePdiError::Internal(e) => Display::fmt(e, f)
+        }
+    }
+}
+
+impl<T: Display> Debug for ConfigureSubdevicePdiError<T> {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        match self {
+            ConfigureSubdevicePdiError::Callback(e) => e.fmt(f),
+            ConfigureSubdevicePdiError::Internal(e) => Debug::fmt(e,f)
+        }
+    }
+}
+
+
+#[cfg(feature = "std")]
+impl<T: std::error::Error> std::error::Error for ConfigureSubdevicePdiError<T> {}
+
+
+impl<T> From<Error> for ConfigureSubdevicePdiError<T> {
+    fn from(value: Error) -> Self {
+        ConfigureSubdevicePdiError::Internal(value)
+    }
+}
+
+impl From<ConfigureSubdevicePdiError<Error>> for Error {
+    fn from(value: ConfigureSubdevicePdiError<Error>) -> Self {
+        match value {
+            ConfigureSubdevicePdiError::Callback(e) => e,
+            ConfigureSubdevicePdiError::Internal(e) => e,
+        }
+    }
+}
+
+impl<T: ConfigureSubdevicePdi, R: DerefMut<Target = T>> ConfigureSubdevicePdi for R {
+    type Error = T::Error;
+
+    async fn configure<'a, S>(
+            &mut self,
+            subdevice: &mut SubDeviceRef<'a, S>,
+            sync_managers: &[SyncManager],
+            fmmu_usage: &[FmmuUsage],
+            direction: PdoDirection,
+            global_offset: &mut PdiOffset,
+        ) -> Result<PdiSegment, Self::Error>
+        where
+            S: DerefMut<Target = SubDevice> {
+        self.deref_mut().configure(subdevice, sync_managers, fmmu_usage, direction, global_offset).await
+    }
+}
+
+/// Configure PDI using the PDO assignments in CoE registers.
+#[derive(Clone, Copy)]
+pub struct ConfigureSubdevicePdiBasedOnCoe;
+impl ConfigureSubdevicePdi for ConfigureSubdevicePdiBasedOnCoe {
```

Comment:
These implementations should be unchanged; I'm just exposing them as separate callbacks. This would have been useful when I was first debugging and wanted to try EEPROM-based setup even though my device supports CoE.